desktop/core/src/desktop/lib/fs/gc/gsconnection.py

```diff
@@ -0,0 +1,152 @@
+
+  Then fill-up the boto HttpRequest with the presigned Url data and lets boto executes the request as usual,
+  so that we get the XML unmarshalling for free.
+
+  Flow:
+    1. signed_url = self.get_signed_url(/bucket/dir/key)
+    2. request = http_request(signed_url)
+    3. return self._mexe(requests)
+  """
+
+  def make_request(self, method, bucket='', key='', headers=None, data='',
+                    query_args=None, sender=None, override_num_retries=None,
+                    retry_handler=None):
+
+    if isinstance(bucket, self.bucket_class):
+      bucket = bucket.name
+    if isinstance(key, Key):
+      key = key.name
+
+    path = self.calling_format.build_path_base(bucket, key)
+    LOG.debug('path=%s' % path)
+
+    auth_path = self.calling_format.build_auth_path(bucket, key)
+    LOG.debug('auth_path=%s' % auth_path)
+
+    host = self.calling_format.build_host(self.server_name(), bucket)
+
+    if query_args:
+      # Call to RAZ for getting the signed headers does not expect the 'generation' argument in the query_args
+      # This might be done to make their side RAZ-GS implementation similar to S3
+
+      # Using regex to remove the 'generation' arg and its value from the query string
+      query_args = re.sub(r'&?generation=[^&]*', '', query_args)
+      query_args = query_args.lstrip('&') # Remove any leading '&' if 'generation' is at the beginning
+
+      path += '?' + query_args
+      LOG.debug('path=%s' % path)
+      auth_path += '?' + query_args
+      LOG.debug('auth_path=%s' % auth_path)
+
+    params = {}
+    # GS expects only one type of headers, either all x-amz-* or all x-goog-*, and the signed headers returned from RAZ are of x-amz-* type
+    # So, we are converting all x-goog-* headers to x-amz-* headers before sending the final request to GS from Hue
+    if headers:
+      updated_headers = {'x-amz-' + key[7:]: value for key, value in headers.items() if key.startswith('x-goog-')}
+      headers.update(updated_headers)
+
+      for key in list(headers.keys()):
+        if key.startswith('x-goog-'):
+          del headers[key]
+
+    http_request = self.build_base_http_request(method, path, auth_path, params, headers, data, host)
+
+    # Actual override starts here
+    LOG.debug('http_request: %s, %s, %s, %s, %s, %s, %s' % (method, path, auth_path, params, headers, data, host))
+    LOG.debug('http_request object: %s' % http_request)
+
+    url = 'https://%(host)s%(path)s' % {'host': host, 'path': path}
+
+    # Do not send the xml data for signing for upload operation
+    xml_data = '' if query_args and 'uploadId=' in query_args else data
+
+    raz_headers = self.get_signed_url(action=method, url=url, headers=headers, data=xml_data)
+    LOG.debug('Raz returned those headers: %s' % raz_headers)
+
+    if raz_headers is not None:
+      http_request.headers.update(raz_headers)
+    else:
+      LOG.error('We got back empty header from Raz for the request %s' % http_request)
```

Comment:
Does it make sense to go ahead and make a call to GS even if the raz headers are empty?
What does RAZ do if I am trying to access a public bucket with no policies defined in RAZ?

Reply:
Good point, will be raising an exception here and for S3 too which is very similar. If we don't have the headers, it does not make sense to continue the operation and we can abort it.

Reply:
@Harshg999 What about the case for publicly accessible buckets. Access to such buckets need not rely on policies in RAZ and will therefore return empty headers. Is my understanding correct?